File: packages/javascript/src/raster-crossfade.js
const UINT32_MAX = 4294967295;
const INT32_MAX = 2147483647;
const MAX_SAFE = 9007199254740991;

/** Validation/access error with a stable catalog code. */
export class RasterCrossfadeError extends Error {
  constructor(code) {
    super(code);
    this.name = "RasterCrossfadeError";
    this.code = code;
  }
}

function invalid() { throw new RasterCrossfadeError("INVALID_INPUT"); }

function isPlainObject(value) {
  if (value === null || typeof value !== "object" || Array.isArray(value)) return false;
  const prototype = Object.getPrototypeOf(value);
  return prototype === Object.prototype || prototype === null;
}

function isPlainArray(value) {
  return Array.isArray(value) && Object.getPrototypeOf(value) === Array.prototype;
}

function exactKeys(record, keys) {
  if (!isPlainObject(record)) invalid();
  if (Reflect.ownKeys(record).length !== keys.length) invalid();
  for (const key of keys) if (!Object.prototype.hasOwnProperty.call(record, key)) invalid();
}

function number(value) {
  if (typeof value !== "number" || !Number.isFinite(value)) invalid();
  return value;
}

function dimension(value) {
  const n = number(value);
  if (n < 1 || n !== Math.floor(n) || n > INT32_MAX) invalid();
  return n;
}

function count(width, height) {
  const product = width * height;
  if (width < 1 || height < 1 || product > INT32_MAX) invalid();
  return product;
}

function unsigned(value) {
  const n = number(value);
  if (n < 0 || n > UINT32_MAX || n !== Math.floor(n)) invalid();
  return n >>> 0;
}

function weight(value) {
  const n = number(value);
  if (n < 0 || n > 1) invalid();
  return n;
}

function objectRaster(value) {
  exactKeys(value, ["width", "height", "pixels"]);
  const width = dimension(value.width);
  const height = dimension(value.height);
  const pixelCount = count(width, height);
  return { width, height, pixels: objectPixels(value.pixels, pixelCount) };
}

function objectPixels(value, pixelCount) {
  if (!isPlainArray(value) || value.length !== pixelCount) invalid();
  const pixels = new Array(pixelCount);
  for (let i = 0; i < pixelCount; i += 1) pixels[i] = unsigned(value[i]);
  return pixels;
}

function quantize(value) {
  const clampedLow = Math.max(0, value);
  const clamped = Math.min(255, clampedLow);
  return Math.floor(clamped + 0.5);
}

function channel(first, second, u, v, a) {
  const p = first * u, q = second * v;
  const sum = p + q;
  const c = sum / a;
  return quantize(c);
}

function composite(first, second, w) {
  if (w === 0) return first;
  if (w === 1) return second;
  const t = 1 - w;
  const firstAlpha = (first >>> 24) & 0xff;
  const af = firstAlpha / 255;
  const secondAlpha = (second >>> 24) & 0xff;
  const as = secondAlpha / 255;
  const u = af * t;
  const v = as * w;
  const a = u + v;
  if (a === 0) return 0;
  const alpha = quantize(a * 255);
  if (alpha === 0) return 0;
  const red = channel((first >>> 16) & 0xff, (second >>> 16) & 0xff, u, v, a);
  const green = channel((first >>> 8) & 0xff, (second >>> 8) & 0xff, u, v, a);
  const blue = channel(first & 0xff, second & 0xff, u, v, a);
  return ((alpha << 24) | (red << 16) | (green << 8) | blue) >>> 0;
}

class Crossfade {
  #width; #height; #pixels;
  constructor(width, height, pixels) {
    this.#width = width; this.#height = height; this.#pixels = pixels;
    Object.freeze(this);
  }
  get width() { return this.#width; }
  get height() { return this.#height; }
  #index(value) {
    if (typeof value !== "number" || !Number.isFinite(value) || value < 0 ||
        value !== Math.floor(value) || value > MAX_SAFE) {
      throw new RasterCrossfadeError("INVALID_INDEX");
    }
    if (value >= this.#pixels.length) throw new RasterCrossfadeError("INDEX_OUT_OF_RANGE");
    return value;
  }
  pixelAt(value) { return this.#pixels[this.#index(value)]; }
  pixels() { return this.#pixels.slice(); }
  toValues() { return { width: this.#width, height: this.#height, pixels: this.#pixels.slice() }; }
}

function validated(width, height, first, second, weights) {
  const pixelCount = count(width, height);
  const output = new Array(pixelCount);
  for (let i = 0; i < pixelCount; i += 1) output[i] = composite(first[i], second[i], weights[i]);
  return new Crossfade(width, height, output);
}

/**
 * Crossfade two same-sized straight ARGB8 rasters using explicit per-pixel weights and
 * premultiplied working channels. Implements raster.crossfade-2d 0.1.0.
 *
 * Provenance: 2017/Generativos/Eyes/eyes002 image stamps; spatial weighting and this
 * portable pixel operation are independently specified project composition work.
 */
export function rasterCrossfade2D(input) {
  exactKeys(input, ["first", "second", "weights"]);
  const first = objectRaster(input.first);
  if (!isPlainObject(input.second)) invalid();
  exactKeys(input.second, ["width", "height", "pixels"]);
  const secondWidth = dimension(input.second.width);
  const secondHeight = dimension(input.second.height);
  const secondCount = count(secondWidth, secondHeight);
  if (first.width !== secondWidth || first.height !== secondHeight) invalid();
  const second = objectPixels(input.second.pixels, secondCount);

  if (!isPlainArray(input.weights) || input.weights.length !== first.pixels.length) invalid();
  const weights = new Array(first.pixels.length);
  for (let i = 0; i < weights.length; i += 1) weights[i] = weight(input.weights[i]);
  return validated(first.width, first.height, first.pixels, second, weights);
}

File: packages/python/procedurals/raster_crossfade.py
"""Owned straight-ARGB8 crossfading with caller-supplied second-input weights.

Implements raster.crossfade-2d 0.1.0. Provenance: 2017/Generativos/Eyes/eyes002 image
stamps; spatial weighting and this portable pixel operation are independently specified
project composition work. Working RGB is encoded and premultiplied only for the stated
weighted arithmetic.
"""
import math

__all__ = ["RasterCrossfadeError", "raster_crossfade_2d"]

_UINT32_MAX = 4294967295
_INT32_MAX = 2147483647
_MAX_SAFE = 9007199254740991


class RasterCrossfadeError(ValueError):
    """Stable catalog code for raster.crossfade-2d."""
    def __init__(self, code):
        super().__init__(code)
        self.code = code


def _invalid():
    raise RasterCrossfadeError("INVALID_INPUT")


def _exact_keys(record, keys):
    if type(record) is not dict or len(record) != len(keys) or any(key not in record for key in keys):
        _invalid()


def _number(value):
    if type(value) not in (int, float) or isinstance(value, bool):
        _invalid()
    value = float(value)
    if not math.isfinite(value):
        _invalid()
    return value


def _dimension(value):
    n = _number(value)
    if n < 1 or n != math.floor(n) or n > _INT32_MAX:
        _invalid()
    return int(n)


def _count(width, height):
    product = width * height
    if width < 1 or height < 1 or product > _INT32_MAX:
        _invalid()
    return product


def _unsigned(value):
    n = _number(value)
    if n < 0 or n > _UINT32_MAX or n != math.floor(n):
        _invalid()
    return int(n)


def _weight(value):
    n = _number(value)
    if n < 0 or n > 1:
        _invalid()
    return n


def _object_pixels(value, pixel_count):
    if type(value) is not list or len(value) != pixel_count:
        _invalid()
    return [_unsigned(pixel) for pixel in value]


def _object_raster(value):
    _exact_keys(value, ("width", "height", "pixels"))
    width = _dimension(value["width"])
    height = _dimension(value["height"])
    pixel_count = _count(width, height)
    return {"width": width, "height": height, "pixels": _object_pixels(value["pixels"], pixel_count)}


def _quantize(value):
    clamped_low = max(0.0, value)
    clamped = min(255.0, clamped_low)
    return math.floor(clamped + 0.5)


def _channel(first, second, u, v, a):
    p = first * u
    q = second * v
    total = p + q
    color = total / a
    return _quantize(color)


def _composite(first, second, weight):
    if weight == 0.0:
        return first
    if weight == 1.0:
        return second
    t = 1.0 - weight
    first_alpha = (first >> 24) & 0xFF
    af = first_alpha / 255.0
    second_alpha = (second >> 24) & 0xFF
    a_s = second_alpha / 255.0
    u = af * t
    v = a_s * weight
    a = u + v
    if a == 0.0:
        return 0
    alpha = _quantize(a * 255.0)
    if alpha == 0:
        return 0
    red = _channel((first >> 16) & 0xFF, (second >> 16) & 0xFF, u, v, a)
    green = _channel((first >> 8) & 0xFF, (second >> 8) & 0xFF, u, v, a)
    blue = _channel(first & 0xFF, second & 0xFF, u, v, a)
    return (alpha << 24) | (red << 16) | (green << 8) | blue


class _Crossfade:
    __slots__ = ("_height", "_pixels", "_width")

    def __init__(self, width, height, pixels):
        self._width = width
        self._height = height
        self._pixels = pixels

    @property
    def width(self):
        return self._width

    @property
    def height(self):
        return self._height

    def _index(self, value):
        if type(value) not in (int, float) or isinstance(value, bool):
            raise RasterCrossfadeError("INVALID_INDEX")
        value = float(value)
        if not math.isfinite(value) or value < 0 or value != math.floor(value) or value > _MAX_SAFE:
            raise RasterCrossfadeError("INVALID_INDEX")
        value = int(value)
        if value >= len(self._pixels):
            raise RasterCrossfadeError("INDEX_OUT_OF_RANGE")
        return value

    def pixel_at(self, value):
        return self._pixels[self._index(value)]

    def pixels(self):
        return list(self._pixels)

    def to_values(self):
        return {"width": self._width, "height": self._height, "pixels": list(self._pixels)}


def _validated(width, height, first, second, weights):
    pixel_count = _count(width, height)
    output = [_composite(first[i], second[i], weights[i]) for i in range(pixel_count)]
    return _Crossfade(width, height, output)


def raster_crossfade_2d(input_value):
    """Crossfade exact portable first, second, and weights records."""
    _exact_keys(input_value, ("first", "second", "weights"))
    first = _object_raster(input_value["first"])
    second_value = input_value["second"]
    _exact_keys(second_value, ("width", "height", "pixels"))
    second_width = _dimension(second_value["width"])
    second_height = _dimension(second_value["height"])
    second_count = _count(second_width, second_height)
    if first["width"] != second_width or first["height"] != second_height:
        _invalid()
    second = _object_pixels(second_value["pixels"], second_count)

    supplied_weights = input_value["weights"]
    if type(supplied_weights) is not list or len(supplied_weights) != len(first["pixels"]):
        _invalid()
    weights = [_weight(value) for value in supplied_weights]
    return _validated(first["width"], first["height"], first["pixels"], second, weights)
